File: habit-tracker/src/main/java/com/habit/habit_tracker/service/TaskService.java
package com.habit.habit_tracker.service;

import java.util.List;

import org.springframework.http.HttpStatus;
import org.springframework.stereotype.Service;

import com.habit.habit_tracker.domain.Task;
import com.habit.habit_tracker.domain.User;
import com.habit.habit_tracker.dto.task.TaskCreateRequest;
import com.habit.habit_tracker.dto.task.TaskUpdateRequest;
import com.habit.habit_tracker.exception.ApiRequestException;
import com.habit.habit_tracker.repository.TaskRepository;
import com.habit.habit_tracker.repository.UserRepository;
import com.habit.habit_tracker.security.AuthUtil;

import jakarta.transaction.Transactional;

import static com.habit.habit_tracker.constants.ErrorMessage.*;

@Service
public class TaskService {
    private final TaskRepository taskRepository;
    private final UserRepository userRepository;
    private final AuthUtil authUtil;

    public TaskService(TaskRepository taskRepository, UserRepository userRepository, AuthUtil authUtil) {
        this.taskRepository = taskRepository;
        this.userRepository = userRepository;
        this.authUtil = authUtil;
    }

    @Transactional
    public Task createTask(TaskCreateRequest taskCreateRequest) {
        User user = authUtil.getAuthenticatedUser();

        user = userRepository.findById(user.getId())
                .orElseThrow(() -> new ApiRequestException(USER_NOT_FOUND, HttpStatus.NOT_FOUND));

        Task task = new Task();
        task.setName(taskCreateRequest.getName());
        task.setDescription(taskCreateRequest.getDescription());
        task.setUser(user);
        task.setDifficulty(taskCreateRequest.getDifficulty());
        task.setDueDate(taskCreateRequest.getDueDate());

        return taskRepository.save(task);
    }

    @Transactional
    public Task updateTask(Long taskId, TaskUpdateRequest taskUpdateRequest) {
        User user = authUtil.getAuthenticatedUser();
        Task task = taskRepository.findByIdAndUserId(taskId, user.getId())
                .orElseThrow(() -> new ApiRequestException(TASK_NOT_FOUND, HttpStatus.NOT_FOUND));

        if (taskUpdateRequest.getName() != null) {
            task.setName(taskUpdateRequest.getName());
        }
        if (taskUpdateRequest.getDescription() != null) {
            task.setDescription(taskUpdateRequest.getDescription());
        }
        if (taskUpdateRequest.getDifficulty() != null) {
            task.setDifficulty(taskUpdateRequest.getDifficulty());
        }
        if (taskUpdateRequest.getStatus() != null) {
            task.setStatus(taskUpdateRequest.getStatus());
        }

        return taskRepository.save(task);
    }

    public Task getTask(Long taskId) {
        User user = authUtil.getAuthenticatedUser();
        return taskRepository.findByIdAndUserId(taskId, user.getId())
                .orElseThrow(() -> new ApiRequestException(TASK_NOT_FOUND, HttpStatus.NOT_FOUND));
    }

    public List<Task> getAllTasks() {
        User user = authUtil.getAuthenticatedUser();
        return taskRepository.findAllByUserId(user.getId());
    }
}

File: habit-tracker/src/main/kotlin/com/habit/habit_tracker/service/TaskService.kt
package com.habit.habit_tracker.service

import com.habit.habit_tracker.constants.ErrorMessage.USER_NOT_FOUND
import com.habit.habit_tracker.constants.ErrorMessage.TASK_NOT_FOUND

import org.springframework.http.HttpStatus
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

import com.habit.habit_tracker.domain.Task
import com.habit.habit_tracker.dto.task.request.TaskCreateRequest
import com.habit.habit_tracker.dto.task.request.TaskUpdateRequest
import com.habit.habit_tracker.exception.ApiRequestException
import com.habit.habit_tracker.repository.TaskRepository
import com.habit.habit_tracker.repository.UserRepository
import com.habit.habit_tracker.security.AuthUtil

@Service
class TaskService(
    private val taskRepository: TaskRepository,
    private val userRepository: UserRepository,
    private val authUtil: AuthUtil
) {
    fun createTask(request: TaskCreateRequest): Task {
        val user = authUtil.getAuthenticatedUser()
            .let { userRepository.findById(it.id!!).orElseThrow {
                ApiRequestException(USER_NOT_FOUND, HttpStatus.NOT_FOUND)
            } } // This is necessary for now, due to detached entity issues :<

        return taskRepository.save(
            Task(
                name = request.name,
                description = request.description,
                user = user,
                dueDate = request.dueDate,
                difficulty = request.difficulty
            )
        )
    }

    @Transactional
    fun updateTask(taskId: Long, request: TaskUpdateRequest): Task {
        val user = authUtil.getAuthenticatedUser()
        val task = taskRepository.findByIdAndUserId(taskId, user.id!!)
            .orElseThrow { ApiRequestException(TASK_NOT_FOUND, HttpStatus.NOT_FOUND) }

        task.apply {
            request.name?.let { name = it }
            request.description?.let { description = it }
            request.difficulty?.let { difficulty = it }
            request.status?.let { status = it }
        }

        return taskRepository.save(task)
    }

    fun getTask(taskId: Long): Task {
        val user = authUtil.getAuthenticatedUser()
        return taskRepository.findByIdAndUserId(taskId, user.id!!)
            .orElseThrow { ApiRequestException(TASK_NOT_FOUND, HttpStatus.NOT_FOUND)}
    }

    fun getAllTasks(): List<Task> {
        val user = authUtil.getAuthenticatedUser()
        return taskRepository.findAllByUserId(user.id!!)
    }
}